hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/file/OMRecoverLeaseRequest.java

```diff
@@ -0,0 +1,264 @@
+                    volumeName, bucketName);
+
+            validateBucketAndVolume(omMetadataManager, volumeName, bucketName);
+
+            String openKeyEntryName = doWork(ozoneManager, recoverLeaseRequest,
+                transactionLogIndex);
+
+            // Prepare response
+            boolean responseCode = true;
+            omResponse.setRecoverLeaseResponse(RecoverLeaseResponse.newBuilder()
+                            .setResponse(responseCode).build())
+                    .setCmdType(RecoverLease);
+            omClientResponse = new OMRecoverLeaseResponse(omResponse.build(),
+                    getBucketLayout(), openKeyEntryName);
+            omMetrics.incNumRecoverLease();
+            LOG.debug("Key recovered. Volume:{}, Bucket:{}, Key:{}", volumeName,
+                bucketName, keyName);
+        } catch (IOException ex) {
+            LOG.error("Fail for recovering lease. Volume:{}, Bucket:{}, Key:{}",
+                volumeName, bucketName, keyName, ex);
+            exception = ex;
+            omMetrics.incNumRecoverLeaseFails();
+            omResponse.setCmdType(RecoverLease);
+            omClientResponse = new OMRecoverLeaseResponse(
+                createErrorOMResponse(omResponse, ex), getBucketLayout(), null);
+        } finally {
+            if (omClientResponse != null) {
+                omClientResponse.setFlushFuture(
+                        ozoneManagerDoubleBufferHelper.add(omClientResponse,
+                                transactionLogIndex));
+            }
+            if (acquiredLock) {
+                omMetadataManager.getLock().releaseWriteLock(BUCKET_LOCK, volumeName,
+                        bucketName);
+            }
+        }
+
+        // Audit Log outside the lock
+        auditLog(ozoneManager.getAuditLogger(), buildAuditMessage(
+                OMAction.RECOVER_LEASE, auditMap, exception,
+                getOmRequest().getUserInfo()));
+
+        return omClientResponse;
+    }
+
+    private String doWork(OzoneManager ozoneManager,
+        RecoverLeaseRequest recoverLeaseRequest, long transactionLogIndex)
+        throws IOException {
+
+        //String dbOzoneKey =
+        //    omMetadataManager.getOzoneKey(volumeName, bucketName,
+        //        keyName);
+        final long volumeId = omMetadataManager.getVolumeId(volumeName);
+        final long bucketId = omMetadataManager.getBucketId(
+            volumeName, bucketName);
+        Iterator<Path> pathComponents = Paths.get(keyName).iterator();
+        long parentID = OMFileRequest.getParentID(volumeId, bucketId,
+            pathComponents, keyName, omMetadataManager,
+            "Cannot recover file : " + keyName
+                + " as parent directory doesn't exist");
+        String fileName = OzoneFSUtils.getFileName(keyName);
+        String dbFileKey = omMetadataManager.getOzonePathKey(volumeId, bucketId,
+            parentID, fileName);
+
+        OmKeyInfo keyInfo = getKey(dbFileKey);
```

Comment:
As my understanding, if a key is present in openKeyTable, entry will not be there in keyTable, please recheck

Reply:
once hsync is called on a file, the file will present in both tables.